### PR TITLE
Partial handling for restrictive annotations on varargs in unannotated code

### PR DESCRIPTION
**Pull request by msridhar**
Partially addresses #1027.  Most importantly, handles the case with unannotated methods and _no_ restrictive annotations; we were getting warnings in some cases with the main branch, which impacts backward compatibility.

I started implementing full support for restrictive annotations on varargs parameters and then realized it would take some refactoring (since there can be a restrictive annotation on the varargs array itself or on its contents, and we have no way to express this in the current code structure).  This PR contains some incomplete code towards that change, which I figured does no harm so I left it in for now.  If desired I can remove it and put it in a separate PR.

### CODE REVIEW

File: nullaway/src/test/java/com/uber/nullaway/LegacyVarargsTests.java [hunk 1]
```diff
@@ -200,7 +200,8 @@ public void testSkipJetbrainsNotNullOnVarArgsFromThirdPartyJars() {
             "    ThirdParty.takesNullableVarargs(o1);", // Empty var args passed
             "    ThirdParty.takesNullableVarargs(o1, o4);",
             "    ThirdParty.takesNullableVarargs(o1, (Object) null);",
-            "    // BUG: Diagnostic contains: passing @Nullable parameter '(Object[]) null' where @NonNull",
+            "    // No error: we require a type-use annotation for nullability of the varargs array itself",
+            // TODO should we preserve the old behavior in legacy mode?
```
Comment (msridhar) [inline nullaway/src/test/java/com/uber/nullaway/LegacyVarargsTests.java:204]: @lazaroclapp with this change, we no longer detect an error when passing a `null` array in the varargs position to a third-party method with a JetBrains `@NotNull` annotation on the varargs parameter.  On the one hand, getting an error here is the "desired" behavior, in that the annotation is supposed to apply to the array itself (see #720).  But, everywhere else, a declaration annotation on the varargs argument is treated as applying to the contents of the array, not the array itself.  To me, adding the extra code complexity to detect these errors in legacy mode for this corner case is not worth it; hopefully Kotlinc will be convinced to put annotations in the "right" place soon.  But I'd be interested in your opinion.

Comment (lazaroclapp) [reply nullaway/src/test/java/com/uber/nullaway/LegacyVarargsTests.java:204]: I think without a coordinated change on `kotlinc`, I would prefer to keep the hack. But I don't feel super strongly about it. It's just an additional test of (a) is the annotation in bytecode, (b) is it the jetbrains annotation, (c) is it in var args... if all that is true, hacky override with a comment linking to the issue. No? Would this be hard to maintain?

Unfortunately, Java code interfacing with Kotlin code is a big portion of what we will see in Android for the foreseeable future, so `kotlinc` quirks might be the kind of thing we can't ignore for the sake of uniformity. Though if you can convince them to generate the "right" bytecode, I'd be happy to see this hack disappear (hopefully we don't need then to analyze magic numbers to see which `kotlinc` version produced the jar)

File: nullaway/src/test/java/com/uber/nullaway/VarargsTests.java [hunk 2]
```diff
@@ -417,4 +417,83 @@ public void varargsPassArrayAndOtherArgs() {
...
+            "    Unannotated.takesVarargsDeclaration(y);",
+            "  }",
+            "  public void testTypeUseOnArray() {",
+            "    Object x = null;",
+            "    Object[] y = null;",
+            "    Unannotated.takesVarargsTypeUseOnArray(x);",
+            // TODO report an error here; will require some refactoring of restrictive annotation
+            //  handling
+            "    Unannotated.takesVarargsTypeUseOnArray(y);",
+            "  }",
```
Comment (lazaroclapp) [inline nullaway/src/test/java/com/uber/nullaway/VarargsTests.java:495]: Should we add at least the test cases for `takesVarargsTypeUseOnElements(...)`? Seems like they would be illustrative.

**Review by lazaroclapp — APPROVED** [review]
Definitely an incomplete fix, as the description clearly states, but also a step in the right direction. See comments below. All addressable in follow ups, of course.

File: nullaway/src/test/java/com/uber/nullaway/VarargsTests.java (hunk 2 above)
Comment (msridhar) [reply nullaway/src/test/java/com/uber/nullaway/VarargsTests.java:507]: Yes, and this case actually works! 🙂 Fixed in 118dc17cf5b329fa58a99cfcaabf70df30b16d48

**Review by msridhar — COMMENTED** [review]
Thanks for the review!
